File: framework/lib/checkbox.dart
part of uxml;

/**
 * Implements a button control that toggles a checked state when pressed.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class CheckBox extends Button {
  /** IsChecked property definition */
  static PropertyDefinition isCheckedProperty;
  static ElementDef checkboxElementDef;
  bool _isCheckedValueForMouseDown = false;

  CheckBox() : super() {
  }

  /**
   * Sets/returns checked state.
   */
  bool get isChecked => getProperty(isCheckedProperty);

  set isChecked(bool value) {
    return setProperty(isCheckedProperty, value);
  }

  /** Overrides UIElement.onMouseDown. */
  void onMouseDown(MouseEventArgs mouseArgs) {
    _isCheckedValueForMouseDown = isChecked;
    super.onMouseDown(mouseArgs);
  }

  /** Overrides Button.onClick to toggle isChecked state. */
  void onClick(EventArgs e) {
    bool oldValue = isChecked;
    if (oldValue != _isCheckedValueForMouseDown) {
      // If a binding updated isChecked during mouseDown, don't perform
      // toggle. The user's intention is to switch from value prior to
      // mouse down to the inverse. This makes sure we don't toggle
      // again and revert isChecked.
      return;
    }
    isChecked = !isChecked;
    if (isChecked != oldValue) { // Check if isChecked was cancelled when set.
      super.notifyListeners(Button.clickEvent, new EventArgs(this));
    }
    e.handled = true;
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => checkboxElementDef;

  /** Registers component. */
  static void registerCheckBox() {
    CheckBox.isCheckedProperty = ElementRegistry.registerProperty("IsChecked",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    CheckBox.checkboxElementDef = ElementRegistry.register("CheckBox",
        Button.buttonElementDef, [isCheckedProperty], null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/CheckBox.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.MouseEventArgs;

import java.util.EnumSet;

/**
 * Implements a button control that toggles a checked state when pressed.
 *
 * @author ferhat
 */
public class CheckBox extends Button {

  /** IsChecked property definition */
  public static PropertyDefinition isCheckedPropDef = PropertySystem.register("IsChecked",
      Boolean.class, CheckBox.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  /**
   * Returns whether button is in checked state.
   */
  public boolean getIsChecked() {
    return (Boolean) getProperty(isCheckedPropDef);
  }

  /**
   * Sets button checked state.
   */
  public void setIsChecked(boolean value) {
    setProperty(isCheckedPropDef, value);
  }

  @Override
  protected void onMouseClick(MouseEventArgs e) {
    setIsChecked(!getIsChecked());
    super.onMouseClick(e);
  }
}
